### DISCUSSION TITLE
IPv6 pools and prefix delegation

### DISCUSSION THREAD

**Discussion by Kentzo** [strongswan/strongswan#1754] [Q&A]
I would like to assign my IPsec roadwarriors an IPv6 virtual IP using the dedicated prefix that can be provided by an upstream router.

Currently see 3 approaches:

1. A external script that uses a template swanctl configuration to dynamically update the in-memory backend configuration
2. An external script that manages the SQL database
3. The eap-radius plugin

Ideally I would rather point strongswan to my DHCPv6, since that seems to be the simplest in terms of maintenance.

**Comment by tobiasbrunner** [comment]
> Ideally I would rather point strongswan to my DHCPv6, since that seems to be the simplest in terms of maintenance.

The _dhcp_ plugin currently only supports IPv4.

**Comment by Kentzo** [comment]
When a pool is reloaded by swanctl and it ends up to be different, what will happen to existing connections that use this pool?

**Reply by tobiasbrunner** [reply]
You currently can't change the addresses of a pool (or unload it) while there are online leases. So you'd have to create a new pool (i.e. different name) and reference that in the config so it gets used for new connections. Existing connections won't be affected by this.

**Reply by Kentzo** [reply]
By can’t you mean the swanctl will fail exit with non-0 or that the change won’t take effect until all current leases are freed?

**Reply by tobiasbrunner** [reply]
Yes, it will fail.

**Reply by Kentzo** [reply]
Would you recommend to just stop the daemon, update the pool and then restart it? Is there an alternative series of swanctl commands that would lead to the same result but more gracefully?

**Reply by tobiasbrunner** [reply]
You could maybe rename the pool and reference the new name in the connection config.

**Reply by Kentzo** [reply]
Can I remove a pool from the connection config without removing pool itself while there are leases that utilize it?

**Reply by tobiasbrunner** [reply]
Sure, the connection config doesn't care what pool is referenced (existing connections are not affected by such a change, they continue to reference the old pool e.g. when releasing virtual IPs once the old SAs are terminated).

**Reply by Kentzo** [reply]
I understand that strongSwan doesn't support it, but in principle with respect to the standard, is it possible for an IKEv2 responder to advertise new virtual IP without breaking the connection?

**Reply by tobiasbrunner** [reply]
Not really. One problem is that the traffic selectors of the CHILD_SA wouldn't match (if they are narrowed to the virtual IP). Since they technically shouldn't change during a rekeying, a new CHILD_SA would have to be created from scratch that's narrowed to the new virtual IP. However, that also would require negotiating the new virtual IP with an INFORMATIONAL exchange first, which, while technically allowed by the protocol, isn't widely implemented (there is also no way to invalidate the previous virtual IP, so the new one would only be an addition). Also, there is no way to notify the client that it has to request an additional virtual IP and create a new CHILD_SA. And while initiating such an exchange from the server might technically be possible, as there are payloads specified in IKEv2 to push configuration attributes, I know of no implementation that actually supports that (the RFC actually says: "There are currently no defined uses for the CFG_SET/CFG_ACK exchange"). So the only real way to handle this is basically terminate the IKE_SA and let the client recreate it (or maybe you could NAT the old virtual IPs to the new prefix until all IKE_SAs that use the old one are gone).

**Reply by Kentzo** [reply]
So dhcp plugin, if implemented to support IPv6, would only work for the initial exchange?

NAT is a suboptimal choice for IPv6 in this case because if you’re going to do any kind of translation from GUA (dynamic but globally-scoped) to ULA (stable but site-scoped), then due to scope preferences in the address selection algorithms this may alter the host behavior. Such as that it will prefer IPv4 over IPv6 if dual-stacked.

**Reply by tobiasbrunner** [reply]
> So dhcp plugin, if implemented to support IPv6, would only work for the initial exchange?

Yes, virtual IPs are only negotiated during the IKE_AUTH exchange.

> NAT is a suboptimal choice for IPv6 in this case because if you’re going to do any kind of translation from GUA (dynamic but globally-scoped) to ULA (stable but site-scoped), then due to scope preferences in the address selection algorithms this may alter the host behavior.

What I suggested would basically just map IPs from the old prefix to the new one while there still are such virtual IPs around, same scope.

**Reply by Kentzo** [reply]
Ah, that might be viable behavior for the dhcp plugin using prefix delegation.

**Reply by tobiasbrunner** [reply]
Why the _dhcp_ plugin? The NAT has to be in the kernel to handle decrypted packets with old virtual IPs.

**Reply by Kentzo** [reply]
Since prefix delegation would be done by the plugin, I expect it to manage necessary facilities in the kernel as well.
